astropy/units/physical.py

```diff
@@ -1,133 +1,423 @@
+    'length'
+
+    Occasionally, a set of units will correspond to more than one
+    physical type.
+
+    >>> u.Pa.physical_type
+    {'energy density', 'pressure', 'stress'}
+    >>> u.Pa.physical_type == 'pressure'
+    True
+    >>> 'energy density' in u.Pa.physical_type
+    True
+
+    Physical types may be used for dimensional analysis.
+
+    >>> length = u.pc.physical_type
+    >>> area = (u.cm ** 2).physical_type
+    >>> length * area
+    'volume'
+
+    Multiplcation, division, and exponentiation work with physical types.
+
+    >>> length * area
+    'volume'
+    >>> area / length
+    'length'
+    >>> length ** 3
+    'volume'
+
+    Unknown physical types will be labelled as ``"unknown"``.
+
+    >>> (u.s ** 13).physical_type
+    'unknown'
+
+    Dimensional analysis may be performed for unknown physical types too.
+
+    >>> length_to_19th_power = (u.m ** 19).physical_type
+    >>> length_to_20th_power = (u.m ** 20).physical_type
+    >>> length_to_19th_power, length_to_20th_power
+    ('unknown', 'unknown')
+    >>> length_to_20th_power / length_to_19th_power
+    'length'
+
+    A string representation of a physical type can include underscores
+    instead of spaces.
+    """
+
+    @staticmethod
+    def _standardize_physical_type_names(physical_type_input):
+        """
+        Convert a string or `set` of strings into a set containing string
+        representations of physical types.
+
+        The strings provided in ``physical_type_input`` can each contain
+        multiple physical types that are separated by a regular slash or
+        comma.  Underscores will be treated as spaces.
+        """
+
+        if not isinstance(physical_type_input, (set, str)):
+            raise TypeError(
+                f"Expecting a string or a set containing strings that "
+                f"represent a physical type, not {physical_type_input}"
+            )
+
+        if isinstance(physical_type_input, str):
+            physical_type_input = {physical_type_input}
+
+        delimiter_pattern = ",|/"
+        standardized_physical_types = set()
+
+        for ptype_input in physical_type_input:
+            if not isinstance(ptype_input, str):
+                raise TypeError(f"Expecting a string, but got {ptype_input}")
+            input_set = set(re.split(delimiter_pattern, ptype_input))
+            processed_set = {s.strip().replace("_", " ") for s in input_set}
+            standardized_physical_types |= processed_set
+
+        return standardized_physical_types
+
+    def __init__(self, unit, physical_types):
+
+        if not isinstance(unit, core.UnitBase):
+            raise TypeError(f"Expecting a unit, not {unit}.")
+
+        self._unit = _replace_temperatures_with_kelvin(unit)
+        self._physical_type_id = self._unit._get_physical_type_id()
+        self._as_set = self._standardize_physical_type_names(physical_types)
+        self._as_list = sorted(self._as_set)
+
+    def __iter__(self):
```

Comment:
I tried the much simpler
```
def __iter__(self):
    yield from self.as_set
```
which did not seem to need a `__next__`. Was anything wrong with that?